libs/wait-for-ecs-deploy/src/dynamic-resource-provider.ts

```diff
@@ -34,110 +34,77 @@ export const dynamicProvider: pulumi.dynamic.ResourceProvider = {
-                .catch((err) => {
-                    throw new pulumi.RunError(err)
-                })
-
-            pulumi.log.debug(`waitForService: services are stable`)
-
-            const services = await ecs
-                .describeServices({
-                    cluster: inputs.clusterName,
-                    services: [inputs.serviceName],
-                })
-                .promise()
-                .then((result) => result.services)
-                .catch((err) => {
-                    throw new pulumi.RunError(err)
-                })
-
-            if (!services) {
-                throw new pulumi.RunError('No services found!')
-            }
-
-            const failedServices = services.filter((s) =>
-                hasFailed(s, inputs.desiredTaskDef),
-            )
-
-            const failureMessage =
-                failedServices.length > 0
-                    ? `One or more services failed to deploy: ${failedServices.map(
-                          (service) => service.serviceName,
-                      )}`
-                    : ''
-
-            const status = failedServices.length > 0 ? 'FAILED' : 'COMPLETED'
-
-            const result: pulumi.UnwrappedObject<State> = {
-                clusterName: inputs.clusterName,
-                serviceName: inputs.serviceName,
-                desiredTaskDef: inputs.desiredTaskDef,
-                failureMessage,
-                status,
-                timeoutMs,
-            }
-
-            pulumi.log.debug(
-                `waitForService: successful return: ${JSON.stringify(result)}`,
-            )
-
-            return result
-        })().catch((err) => {
-            throw err
-        }),
-    ])
-    pulumi.log.debug(`waitForService: return retval: ${JSON.stringify(retval)}`)
+
+    const ecs = new aws.ECS({
+        region: inputs.awsRegion,
+        credentials: inputs.assumeRole
+            ? new aws.TemporaryCredentials({
+                  RoleArn: inputs.assumeRole,
+                  RoleSessionName: `wait-for-ecs.ecs.${cuid()}`,
+              })
+            : undefined,
+    })
+
+    // current circuit breakers don't catch all error conditions,
+    // eg https://github.com/aws/containers-roadmap/issues/1206 --
+    // this timeout will cause a deployment to fail after a certain
+    // amount of time.
+    await ecs
+        .waitFor('servicesStable', {
+            cluster: inputs.clusterName,
+            services: [inputs.serviceName],
+            $waiter: {
+                delay: 6,
+                maxAttempts: Math.max(1, Math.round(timeoutMs / (1000 * 6))),
+            },
+        })
+        .promise()
+
+    pulumi.log.debug(`waitForService: services are stable`)
+
+    const services = await ecs
+        .describeServices({
+            cluster: inputs.clusterName,
+            services: [inputs.serviceName],
+        })
+        .promise()
+        .then((result) => result.services)
+
+    if (!services) {
```

Comment:
This would never be undefined because you will always have a promise

Reply:
Line 76 does a `.then(result => result.services)` so it has to be checked